### PR TITLE
Add the automatic differentiation multibody solver based on JAX

### PR DESCRIPTION
**Pull request by ben-l-p**
I have added a new solver ```NonlinearDynamicMultibodyJAX``` with the similar functionality as ```NonlinearDynamicMultibody```, except it does not include any ```DynamicTrim``` routine (this can be added in due course when required). This new solver makes use of AD to calculate the Jacobians which leads to a much nicer and more compact definition of constraints, particularly for use of the penalty method. A lot of the key constraint types from the original solver are included here, such as:
 - Free hinge between beams with arbitrary axis line
 - Free hinge between beam and a fixed point in the inertial FoR
 - Spherical joint between beams
 - Spherical joints between beam and a fixed point in the inertial FoR
 - Fully constrained joint between beams
 - Fully constrained beam to a fixed point in the inertial FoR (this is effectively a less efficient way of created a “prescribed” beam, but was useful for testing)

The constraints numerics and the derivatives are defined in the ```sharpy/structure/utils/lagrangeconstraintsjax``` file. I have included new versions of all the other parts of the framework to prevent changes to one breaking the other solver etc., hence why ```sharpy/utils/multibodyjax``` and ```sharpy/solvers/timeintegratorsjax``` exist. 

Also included is the ability for controlled actuation between beams, which I primarily implemented for testing variable sweep wings, however the formulation/implementation is general for any 3D rotation. These can be controlled with the new ```sharpy/controllers/multibodycontroller``` controller type, which takes a Cartesian rotation vector time series as input. To allow this to sweep the wing correctly, the ability to warp the aero grid has been added with a new ```aerogrid_warp_factor``` parameter in the multibody file. This allows for a gradual sweep around a kink, and should not effect for existing cases, as it has no effect is the parameter is not included in the H5 file. 

A test case for this solver is included in the form of a flexible double pendulum comparison. A free double pendulum case is run, the angles from the two hinges extracted and applied onto a prescribed model, where both should yield the “same” result for structural deflections. 

Also included is some documentation on the multibody case files (should be general for both multibody solvers). In recent testing I have found a bug in the ```StaticTrim``` routine which was not present in v2.0 and I am currently investigating, which will be another PR in due course, but it is not connected to the multibody implementation. 

I have also done some code cleanups I found along the way, due to PyCharm largely doing it for me, but for files not related to the new solver the functionality should not have been changed. Lastly, I found one of the unit test cases was creating files which didn't get deleted at the end, which has also been fixed.

### CODE REVIEW

File: sharpy/aero/models/aerogrid.py [hunk 1]
```diff
@@ -270,15 +265,28 @@ def generate_zeta_timestep_info(self, structure_tstep, aero_tstep, beam, setting
                             raise NotImplementedError(str(self.data_dict['control_surface_type'][i_control_surface]) +
                                 ' control surfaces are not yet implemented')
 
-
+                # add sweep for aerogrid warping in constraint defintition
```
Comment (ben-l-p) [inline sharpy/aero/models/aerogrid.py:268]: New code for dynamically sweeping the aero grid. This shouldn't impact any existing cases as it will ignore it if the warp factor parameter does not exist.

File: sharpy/aero/utils/airfoilpolars.py [hunk 2]
```diff
@@ -62,7 +61,7 @@ def get_coefs(self, aoa_deg):
         cd = self.cd_interp(aoa_deg)
         cm = self.cm_interp(aoa_deg)
 
-        return cl, cd, cm
+        return cl[0], cd[0], cm[0]
```
Comment (ben-l-p) [inline sharpy/aero/utils/airfoilpolars.py:64]: Fixed one of the pesky errors converting an array to a scalar during the unit test - this seems to work fine.

File: sharpy/controllers/multibodycontroller.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
+import numpy as np
+import os
+
+import sharpy.utils.controller_interface as controller_interface
+import sharpy.utils.settings as settings
+
+
+@controller_interface.controller
+class MultibodyController(controller_interface.BaseController):
```
Comment (ben-l-p) [inline sharpy/controllers/multibodycontroller.py:9]: New multibody controller for setting angle between beams.

File: sharpy/solvers/nonlineardynamicmultibodyjax.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
+import numpy as np
```
Comment (ben-l-p) [inline sharpy/solvers/nonlineardynamicmultibodyjax.py:1]: New solver, very similar to the existing multibody solver in its inputs.

File: sharpy/solvers/steplinearuvlm.py [hunk 5]
```diff
@@ -323,11 +312,10 @@ def run(self, **kwargs):
         """
 
         aero_tstep = settings_utils.set_value_or_default(kwargs, 'aero_step', self.data.aero.timestep_info[-1])
-        structure_tstep = settings_utils.set_value_or_default(kwargs, 'structural_step', self.data.structure.timestep_info[-1])
-        convect_wake = settings_utils.set_value_or_default(kwargs, 'convect_wake', False)
```
Comment (ben-l-p) [inline sharpy/solvers/steplinearuvlm.py:327]: I cut some parameters here as they aren't actually referenced again

File: sharpy/solvers/timeintegrators.py [hunk 6]
```diff
@@ -71,6 +71,8 @@ class NewmarkBeta(_BaseTimeIntegrator):
 
     def __init__(self):
 
+        self.sys_size = None
```
Comment (ben-l-p) [inline sharpy/solvers/timeintegrators.py:74]: Some variables used later were not declared in __init__

File: sharpy/solvers/timeintegratorsjax.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
+import numpy as np
```
Comment (ben-l-p) [inline sharpy/solvers/timeintegratorsjax.py:1]: New time integrator for the JAX solver, should have the same numerics as the existing implementation.

File: sharpy/structure/models/beam.py [hunk 8]
```diff
@@ -544,7 +476,8 @@ def get_body(self, ibody):
         int_list_nodes = np.arange(0, ibody_beam.num_node, 1)
         for ielem in range(ibody_beam.num_elem):
             for inode_in_elem in range(ibody_beam.num_node_elem):
-                ibody_beam.connectivities[ielem, inode_in_elem] = int_list_nodes[ibody_nodes == ibody_beam.connectivities[ielem, inode_in_elem]]
+                ibody_beam.connectivities[ielem, inode_in_elem] = int_list_nodes[
```
Comment (ben-l-p) [inline sharpy/structure/models/beam.py:479]: Another fix for converting arrays to scalars which creates a warning during unit test.

File: sharpy/structure/utils/lagrangeconstraintsjax.py [hunk 9]
```diff
@@ -0,0 +1,498 @@
+from typing import Callable, Any, Optional, Type, cast
```
Comment (ben-l-p) [inline sharpy/structure/utils/lagrangeconstraintsjax.py:1]: This is where the multibody constraint numerics take place

File: sharpy/utils/generate_cases.py [hunk 10]
```diff
@@ -1989,13 +1991,15 @@ def check(self):
             raise RuntimeError(("'behaviour' parameter is required in '%s' lagrange constraint" % self.behaviour))
 
 
-def generate_multibody_file(list_LagrangeConstraints, list_Bodies, route, case_name):
+def generate_multibody_file(list_LagrangeConstraints, list_Bodies, route, case_name, use_jax=False):
```
Comment (ben-l-p) [inline sharpy/utils/generate_cases.py:1994]: I need to add this additional setting as the generate case class checks for constraints in the existing solver; as I have new constraints which aren't implemented there, it fails when this is False.

File: sharpy/solvers/nonlineardynamicmultibodyjax.py [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+import typing
```
[the remark this answers is not in the document]
Comment (ben-l-p) [reply sharpy/solvers/nonlineardynamicmultibodyjax.py:2]: Typing is the built-in Python type hinting, so this doesn't require any new packages

File: sharpy/solvers/statictrim.py [hunk 12]
```diff
@@ -317,17 +317,12 @@ def trim_algorithm(self):
 
     def evaluate(self, alpha, deflection_gamma, thrust):
         if not np.isfinite(alpha):
-            import pdb; pdb.set_trace()
+            raise ValueError
```
Comment (kccwing) [inline sharpy/solvers/statictrim.py:320]: Here I think the reason behind breaking and not terminating the run is because staticcoupled has no post-processing - so rather than killing it (and leaving no trace to what happened) they preferred to leave a break point here. If we haven't got plans to introduce post-processing for staticcoupled iterations, is there any neater alternative to just throwing an error here?

Comment (ben-l-p) [reply sharpy/solvers/statictrim.py:320]: I was having the trace cause me issues when running on the HPC a while ago, I'm not sure if it's good practice to have traces like this in “production” code (although I may be wrong). This error is thrown when one of the trim gradients is zero, which is currently occurring for some of my cases and I believe is a bug, that I'm currently looking into. If static coupled fails, the code won't get this far, as the FORTRAN will instead throw a singular matrix error.

File: sharpy/structure/models/beam.py (hunk 8 above)
Comment (kccwing) [reply sharpy/structure/models/beam.py:479]: good spot and fix!

File: tests/coupled/multibody/double_slanted_pendulum/test_double_pendulum_slanted.py [hunk 13]
```diff
@@ -322,7 +301,7 @@ def test_doublependulum_hinge_slanted_lateralrot(self):
     def tearDown(self):
         solver_path = os.path.abspath(os.path.dirname(os.path.realpath(__file__)))
         solver_path += '/'
-        for name in [name_hinge_slanted, name_hinge_slanted_pen, name_hinge_slanted_lateralrot]:
```
[the remark this answers is not in the document]
Comment (kccwing) [reply tests/coupled/multibody/double_slanted_pendulum/test_double_pendulum_slanted.py:325]: ahh sorry! my bad😂😂

File: tests/coupled/static/smith_g_2deg/generate_smith_g_2deg.py [hunk 14]
```diff
@@ -208,7 +196,7 @@ def generate_aero_file():
 
     working_elem = 0
     working_node = 0
-    # right wing (surface 0, beam 0)
+
```
Comment (kccwing) [inline tests/coupled/static/smith_g_2deg/generate_smith_g_2deg.py:199]: my take is to leave these in - it helps with understanding the model generation procedure which if now there's no simplification coming up the pipeline seems to have a big learning curve.

File: tests/coupled/static/smith_g_4deg/generate_smith_g_4deg.py [hunk 15]
```diff
@@ -207,7 +193,7 @@ def generate_aero_file():
 
     working_elem = 0
     working_node = 0
-    # right wing (surface 0, beam 0)
+
```
Comment (kccwing) [inline tests/coupled/static/smith_g_4deg/generate_smith_g_4deg.py:196]: likewise, I suggest leaving them in, happy to have a discussion

**Review by kccwing — APPROVED** [review]
Nice work, @ben-l-p! Thanks for building a parallel for most files used by the new jax routines - that should hopefully make the work of migrating over in the future much easier. Happy to merge given this passes all tests and that you've been using it day-to-day already.

File: sharpy/solvers/nonlineardynamicmultibodyjax.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (kccwing) [reply sharpy/solvers/nonlineardynamicmultibodyjax.py:2]: ahhhhhh ok! thanks

File: sharpy/solvers/statictrim.py (hunk 12 above)
Comment (kccwing) [reply sharpy/solvers/statictrim.py:320]: that makes sense - I meant more in the off chance that staticcoupled did return eg. same total forces for a parametrised run with different geometry (perhaps the staticcoupled get total forces function working on the timestep[-1] time information, and somehow got contaminated? just a wild guess) - that would kill the trim routine like what you've seen. but yeah I agree it is probably best to clean up production code for an HPC environment assuming no user input is possible further

**Comment by wong-hl** [comment]
Btw I really like your PR message, it's super clear about what's in the PR etc

File: sharpy/aero/models/aerogrid.py [hunk 16]
```diff
@@ -270,15 +265,28 @@ def generate_zeta_timestep_info(self, structure_tstep, aero_tstep, beam, setting
...
+                    if structure_tstep.mb_prescribed_dict is not None:
+                        for i_constraint in range(structure_tstep.mb_dict['num_constraints']):
+                            try:
+                                cst_name = f"constraint_{i_constraint:02d}"
+                                ctrl_id = structure_tstep.mb_dict[cst_name]['controller_id'].decode('UTF-8')
+                                f_warp = structure_tstep.mb_dict[cst_name]['aerogrid_warp_factor'][i_elem, i_local_node]
+                                ang_z = structure_tstep.mb_prescribed_dict[ctrl_id]['delta_psi'][2]
+                                ang_warp += f_warp * ang_z
+                            except KeyError:
+                                continue
```
[the remark this answers is not in the document]
Comment (ben-l-p) [reply sharpy/aero/models/aerogrid.py:281]: The constraints here are input as dictionaries and have different key-value pairs depending on their functionality (hinge axis, node number etc.). Warping the aerodynamic grid will occur here if a constraint has both ```controller_id``` and ```aerogrid_warp_factor``` entries, with the intended behavior to skip this code if a constraint is missing one. A key error will occur if it's a constraint that is missing either of these entries (and therefore is not a constraint which requires the warping). Of course, it could technically fail if I had coded it wrong and ```delta_psi``` is not defined, and this would incorrectly ignore this code, but I'm pretty sure that can't happen.

File: sharpy/controllers/multibodycontroller.py [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+        "Write a time history of input, required input, " + "and control"
+    )
+
+    settings_table = settings.SettingsTable()
+    __doc__ += settings_table.generate(
+        settings_types, settings_default, settings_description, settings_options
+    )
+
+    def __init__(self):
+        self.in_dict = None
```
[the remark this answers is not in the document]
Comment (ben-l-p) [reply sharpy/controllers/multibodycontroller.py:48]: I have included this to be consistent with the other controllers available, it's the settings dictionary, and I can add a comment to state this. I do not like how the controller code (for all of them) is structured, but I would rather they're all the same at least.

File: sharpy/controllers/multibodycontroller.py [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+        if controlled_state["structural"].mb_prescribed_dict is None:
+            controlled_state["structural"].mb_prescribed_dict = dict()
+        controlled_state["structural"].mb_prescribed_dict[self.controller_id] = {
+            "psi": control_command,
+            "psi_dot": psi_dot,
+        }
+        controlled_state["structural"].mb_prescribed_dict[self.controller_id].update(
+            {"delta_psi": control_command - self.prescribed_ang_time_history[0, :]}
+        )
+
```
[the remark this answers is not in the document]
Comment (ben-l-p) [reply sharpy/controllers/multibodycontroller.py:167]: Good spot!

File: sharpy/solvers/nonlineardynamicmultibodyjax.py [hunk 19]
```diff
@@ -0,0 +1,368 @@
...
+                self.data.structure.timestep_info[-1].mb_FoR_vel[ibody, :] \
+                    = self.data.structure.timestep_info[-1].for_vel
+
+        # find total number of LC equations
+        dict_of_lc = lagrangeconstraints.DICT_OF_LC
+        lc_cls_list: list[typing.Type[lagrangeconstraints.Constraint]] = []
+        lc_settings: list[dict] = []
+        self.num_lm_tot = 0
+        for i in range(self.data.structure.ini_mb_dict['num_constraints']):
+            lc_settings.append(self.data.structure.ini_mb_dict[f'constraint_{i:02d}'])
```
[the remark this answers is not in the document]
Comment (ben-l-p) [reply sharpy/solvers/nonlineardynamicmultibodyjax.py:135]: The number needs to be formatted to have two digits, i.e. 00, 01, 02. Not a fan of this method as it means this has to happen in a few places in the code, but unless if I overhaul the multibody then it's a necessity for backwards compatibility.
